connector/src/websocket_source.rs

```diff
@@ -17,30 +17,165 @@ use std::{
+        );
+    }
+
+    let mut slot_sub = client.slots_updates_subscribe().map_err_anyhow()?;
+
+    // note: the snapshot refresh schedule is local to the feed_data method and will be reset in outer loop
+    let mut last_snapshot = Instant::now().sub(SNAPSHOT_REFRESH_INTERVAL);
+
+    // consume from channels until an error happens
+    loop {
+
+        // occasionally cause a new snapshot to be produced
+        // including the first time
+        if last_snapshot + SNAPSHOT_REFRESH_INTERVAL <= Instant::now() {
+            let snapshot_rpc_http_url = config.snapshot.rpc_http_url.clone();
+            let response =
+                get_snapshot_gma(snapshot_rpc_http_url.clone(), account_ids.clone()).await;
+            let snapshot = response.context("gma snapshot response").map_err_anyhow();
+            match snapshot {
+                Ok(SnapshotMultipleAccounts { snapshot_slot, snapshot_accounts }) => {
+                    debug!(
+                        "fetched new gma snapshot slot={} len={:?} time={:?}",
+                        snapshot_slot,
+                        snapshot_accounts.len(),
+                        Instant::now() - SNAPSHOT_REFRESH_INTERVAL - last_snapshot
+                    );
+                    sender
+                        .send(WebsocketMessage::SnapshotUpdate((snapshot_slot, snapshot_accounts)))
+                        .await
+                        .expect("sending must succeed");
+                }
+                Err(err) => {
+                    warn!("failed to parse snapshot from rpc url {}: {}", snapshot_rpc_http_url.clone(), err);
+                }
+            }
+            last_snapshot = Instant::now();
+        }
+
+        tokio::select! {
+            account = account_subs.next() => {
+                match account {
+                    Some((account_id, response)) => {
+                        sender.send(
+                            WebsocketMessage::SingleUpdate(
+                                response
+                                    .map( |r: Response<UiAccount>|
+                                        Response {
+                                            context: r.context,
+                                            value: RpcKeyedAccount {
+                                                pubkey: account_id.clone(),
+                                                account: r.value }})
+                                    .map_err_anyhow()?)).await.expect("sending must succeed");
+                    },
+                    None => {
+                        // note: this might loop if the filter did not return anything
+                        warn!("account stream closed");
+                        // that is weired - assume that it is okey to cancel in all cases
+                        // if filter_config.program_ids.is_empty() {
+                        //     return Ok(());
+                        // }
+                        return Ok(());
```

Comment:
note: that code is a bit unclear to me; assuming now that it is okey to cancel in all cases